File: core/src/mods/ssl.rs
use std::process::{Command, Stdio};

use crate::mods::logger::{log_step, log_success, log_warning};
use crate::mods::models::DomainConfig;

/// Check if certbot nginx plugin is available
fn check_certbot_nginx_plugin() -> bool {
    if let Ok(output) = Command::new("certbot")
        .args(&["plugins", "--text"])
        .output()
    {
        let plugins_text = String::from_utf8_lossy(&output.stdout);
        plugins_text.contains("nginx") || plugins_text.contains("* nginx")
    } else {
        false
    }
}

/// Install certbot nginx plugin
fn install_certbot_nginx_plugin() -> Result<(), String> {
    log_warning("⚠️  Certbot nginx plugin not found. Installing...");
    
    let output = Command::new("apt-get")
        .args(&["install", "-y", "python3-certbot-nginx"])
        .stdin(Stdio::inherit())
        .stdout(Stdio::inherit())
        .stderr(Stdio::inherit())
        .output()
        .map_err(|e| format!("Failed to install certbot nginx plugin: {}", e))?;
    
    if !output.status.success() {
        return Err("Failed to install python3-certbot-nginx package".to_string());
    }
    
    log_success("✓ Certbot nginx plugin installed");
    Ok(())
}

pub fn setup_ssl(config: &DomainConfig) -> Result<(), String> {
    log_step(&format!("> Setting up SSL for {}...", config.domain));

    let email = config.email.as_ref().ok_or("Email required for SSL")?;

    // Check if nginx plugin is available, install if missing
    if !check_certbot_nginx_plugin() {
        install_certbot_nginx_plugin()?;
    }

    let output = Command::new("certbot")
        .args(&[
            "certonly",
            "--nginx",
            "-d",
            &config.domain,
            "--email",
            email,
            "--agree-tos",
            "--non-interactive",
        ])
        .output()
        .map_err(|e| format!("Failed to run certbot: {}", e))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        let stdout = String::from_utf8_lossy(&output.stdout);
        
        // Check if it's a plugin error and try to fix
        if stderr.contains("does not appear to be installed") || stderr.contains("nginx plugin") {
            log_warning("⚠️  Certbot nginx plugin error detected. Attempting to fix...");
            install_certbot_nginx_plugin()?;
            
            // Retry certbot
            log_step("> Retrying SSL certificate request...");
            let retry_output = Command::new("certbot")
                .args(&[
                    "certonly",
                    "--nginx",
                    "-d",
                    &config.domain,
                    "--email",
                    email,
                    "--agree-tos",
                    "--non-interactive",
                ])
                .output()
                .map_err(|e| format!("Failed to retry certbot: {}", e))?;
            
            if !retry_output.status.success() {
                let retry_stderr = String::from_utf8_lossy(&retry_output.stderr);
                return Err(format!("Certbot failed after plugin installation:\n{}", retry_stderr));
            }
        } else {
            return Err(format!("Certbot failed:\n{}\n{}", stderr, stdout));
        }
    }

    log_success("✓ SSL certificate obtained");
    Ok(())
}

File: core-go/ssl/ssl.go
package ssl

import (
	"fmt"
	"os"
	"os/exec"
	"strings"
	"sync"
	"time"

	"xynginc/logger"
	"xynginc/models"
)

var certbotMutex sync.Mutex

func checkCertbotNginxPlugin() bool {
	cmd := exec.Command("certbot", "plugins", "--text")
	out, err := cmd.CombinedOutput()
	if err == nil {
		pluginsText := string(out)
		return strings.Contains(pluginsText, "nginx") || strings.Contains(pluginsText, "* nginx")
	}
	return false
}

func installCertbotNginxPlugin() error {
	logger.Warning("⚠️  Certbot nginx plugin not found. Installing...")

	cmd := exec.Command("apt-get", "install", "-y", "python3-certbot-nginx")
	cmd.Stdin = os.Stdin
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	err := cmd.Run()
	if err != nil {
		return fmt.Errorf("failed to install python3-certbot-nginx package: %v", err)
	}

	logger.Success("✓ Certbot nginx plugin installed")
	return nil
}

func isCertbotRunning() bool {
	cmd := exec.Command("pgrep", "-f", "certbot")
	err := cmd.Run()
	return err == nil
}

func cleanupStaleLockFiles() {
	lockFiles := []string{
		"/var/log/letsencrypt/certbot.lock",
		"/var/lib/letsencrypt/certbot.lock",
		"/etc/letsencrypt/.certbot.lock",
	}
	for _, lockFile := range lockFiles {
		if _, err := os.Stat(lockFile); err == nil {
			os.Remove(lockFile)
		}
	}
}

func runCertbotWithRetry(args []string) ([]byte, error) {
	maxRetries := 6
	backoff := 3 * time.Second

	for i := 0; i < maxRetries; i++ {
		cmd := exec.Command("certbot", args...)
		out, err := cmd.CombinedOutput()
		if err == nil {
			return out, nil
		}

		stderrText := string(out)

		// Case 1: Plugin missing
		if strings.Contains(stderrText, "does not appear to be installed") || strings.Contains(stderrText, "nginx plugin") {
			logger.Warning("⚠️  Certbot nginx plugin error detected. Attempting to fix...")
			if pErr := installCertbotNginxPlugin(); pErr != nil {
				return nil, pErr
			}
			logger.Step("> Retrying SSL certificate request...")
			retryCmd := exec.Command("certbot", args...)
			return retryCmd.CombinedOutput()
		}

		// Case 2: Another instance running or lock conflict
		if strings.Contains(stderrText, "Another instance of Certbot is already running") ||
			strings.Contains(stderrText, "certbot.lock") ||
			strings.Contains(stderrText, "LockError") {

			if isCertbotRunning() {
				logger.Warning(fmt.Sprintf("⚠️  Certbot lock active (another process running). Waiting %v before retry (%d/%d)...", backoff, i+1, maxRetries))
				time.Sleep(backoff)
				continue
			} else {
				// No certbot process is running, stale lock file!
				logger.Warning("⚠️  Stale Certbot lock file detected with no running process. Cleaning up lock files...")
				cleanupStaleLockFiles()
				time.Sleep(1 * time.Second)
				continue
			}
		}

		return nil, fmt.Errorf("certbot failed:\n%s", stderrText)
	}

	return nil, fmt.Errorf("certbot failed: lock conflict could not be resolved after multiple retries")
}

func SetupSSL(config *models.DomainConfig) error {
	certbotMutex.Lock()
	defer certbotMutex.Unlock()

	logger.Step(fmt.Sprintf("> Setting up SSL for %s...", config.Domain))

	certPath := fmt.Sprintf("/etc/letsencrypt/live/%s/fullchain.pem", config.Domain)
	if _, err := os.Stat(certPath); err == nil {
		logger.Success(fmt.Sprintf("✓ SSL certificate already exists for %s", config.Domain))
		return nil
	}

	if !checkCertbotNginxPlugin() {
		if err := installCertbotNginxPlugin(); err != nil {
			return err
		}
	}

	args := []string{
		"certonly",
		"--nginx",
		"-d", config.Domain,
		"--agree-tos",
		"--non-interactive",
	}

	if config.Email != "" {
		args = append(args, "--email", config.Email)
	} else {
		args = append(args, "--register-unsafely-without-email")
	}

	_, err := runCertbotWithRetry(args)
	if err != nil {
		return err
	}

	logger.Success("✓ SSL certificate obtained")
	return nil
}

